linker/shared/src/test/scala/org/scalajs/linker/frontend/modulesplitter/TaggerTest.scala

```diff
@@ -0,0 +1,434 @@
+        moduleAID -> ListSet(publicClassName)
+      )
+      val info = new DependencyInfo(classDependencies, publicModuleDependencies)
+      val tagger = new Tagger(info, Set.empty)
+      tagger.tagAll(Iterable.empty)
+    }
+
+    val finalModuleMap = locally {
+      val publicModuleDependencies = ListMap(
+        moduleAID -> ListSet(publicClassName),
+        moduleBID -> ListSet(publicClassName)
+      )
+      val info = new DependencyInfo(classDependencies, publicModuleDependencies)
+      val tagger = new Tagger(info, Set.empty)
+      tagger.tagAll(Iterable.empty)
+    }
+
+    val initialStaticModuleID = initialModuleMap(staticClassName)
+
+    val finalStaticModuleID = finalModuleMap(staticClassName)
+    val finalPublicModuleID = finalModuleMap(publicClassName)
+
+    assertNotEquals(initialStaticModuleID, finalStaticModuleID)
+    assertEquals(finalPublicModuleID, finalStaticModuleID)
+  }
+
+  @Test def testModuleIDOfStaticChangesWithNewTransitiveDynamicPathSameModule(): Unit = {
+    // Initially the class has one direct path.
+    // Ensure that adding a dynamic path from the same public modules will change the module ID.
+
+    val moduleAID = ModuleID("modA")
+
+    val publicClassName = ClassName("public")
+    val dynamicClassName = ClassName("dynamic")
+    val staticClassName = ClassName("static")
+
+    val initialModuleMap = locally {
+      val classDependencies = ListMap(
+        publicClassName -> new ClassInfo(ListSet(staticClassName), Set.empty),
+        staticClassName -> new ClassInfo(Set.empty, Set.empty)
+      )
+      val publicModuleDependencies = ListMap(
+        moduleAID -> ListSet(publicClassName)
+      )
+      val info = new DependencyInfo(classDependencies, publicModuleDependencies)
+      val tagger = new Tagger(info, Set.empty)
+      tagger.tagAll(Iterable.empty)
+    }
+
+    val finalModuleMap = locally {
+      val classDependencies = ListMap(
+        publicClassName -> new ClassInfo(ListSet(staticClassName), ListSet(dynamicClassName)),
+        dynamicClassName -> new ClassInfo(ListSet(staticClassName), Set.empty),
+        staticClassName -> new ClassInfo(Set.empty, Set.empty)
+      )
+      val publicModuleDependencies = ListMap(
+        moduleAID -> ListSet(publicClassName)
+      )
+      val info = new DependencyInfo(classDependencies, publicModuleDependencies)
+      val tagger = new Tagger(info, Set.empty)
+      tagger.tagAll(Iterable.empty)
+    }
+
+    val initialStaticModuleID = initialModuleMap(staticClassName)
+
+    val finalPublicModuleID = finalModuleMap(publicClassName)
+    val finalDynamicModuleID = finalModuleMap(dynamicClassName)
+    val finalStaticModuleID = finalModuleMap(staticClassName)
+
+    assertNotEquals(initialStaticModuleID, finalStaticModuleID)
+    assertNotEquals(finalPublicModuleID, finalStaticModuleID)
+    // TODO: Why are there 3 modules and not 2?
+    assertNotEquals(finalDynamicModuleID, finalStaticModuleID)
```

Comment:
Yes:

- The public module with the public class
- The dynamic module with the dynamic class
- The core module with the shared dependencies of the public and the dynamic class (this cannot be in the public module, because it must be imported by the dynamic module).